storage: deflake TestStoreMetrics

tbg opened the pull request:
This was a tough one. Several problems were addressed, all variations on the
same theme:
- DistSenders in multiTestContext use a shared global stopper, but they
  may be called on goroutines which belong to a Store-level task. If that
  Store wants to quiesce and the DistSender can't finish its task because
  that same Store is already in quiescing mode, deadlocks occurred.
  The unfortunate solution is plugging in a channel which draws from two
  Stoppers, one of which may be quiesced and replaced multiple times.
- Additional deadlocks were caused due to multiTestContext's transport,
  which acquired a read lock that was formerly held in write mode throughout
  mtc.stopStore() (circumvented by dropping the lock there while quiescing).
- verifyStats was stopping individual Stores to perform computations without
  moving parts. Stopping individual Stores is tough when their tasks may be
  stuck on other Stores but can't complete while their own Store is already
  quiescing. Instead, verifyStats stops _all stores_ simultaneously, regardless
  of which Store is actively being investigated.

Prior to these changes, failed in a few hundred to a few thousand iters
(depending on how many of the above were partially addressed):

```
$ make stressrace PKG=./storage TESTS=TestStoreMetrics TESTTIMEOUT=10s STRESSFLAGS='-maxfails 1 -stderr -p 128 -timeout 15m'
15784 runs so far, 0 failures, over 8m0s
```

Fixes #7678.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7809)

<!-- Reviewable:end -->

tamird commented:
Partially reviewed, will resume later.

---

Reviewed 4 of 5 files at r1.
Review status: 4 of 5 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/client_metrics_test.go, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KM_UWcq0yBM0rsQmUDI:-KM_UWcq0yBM0rsQmUDJ:-1769119388) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_metrics_test.go#L82)):*
> ```Go
> 	// least when bringing down the nodes jeopardizes majorities.
> 	for i := 0; i < numStores; i++ {
> 		go func(i int) {
> ```

danger zone

---

*[storage/client_metrics_test.go, line 102 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KM_Uob8idq49SNKutxO:-KM_Uob9TmoGDwBmcKTY:271814374) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_metrics_test.go#L102)):*
> ```Go
> 		// Sanity regression check for bug #4624: ensure intent count is zero.
> 		if a := realStats.IntentCount; a != 0 {
> 			fatalf("Expected intent count to be zero, was %d", a)
> ```

since you're here, uncapitalize these

---

*[storage/client_metrics_test.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KM_CuY6NanDiHmTk7L_:-KM_CuY6NanDiHmTk7La:-1879080685) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_metrics_test.go#L132)):*
> ```Go
> 
> 	for i := 0; i < numStores; i++ {
> 		// Restart the store at the provided index.
> ```

this comment is truly worthless

---

*[storage/client_test.go, line 544 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KM_V4BXBwxUYPjgqM5-:-KM_V4BXBwxUYPjgqM50:-1054325866) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L544)):*
> ```Go
> 	if len(m.dbs) <= idx {
> 		retryOpts := base.DefaultRetryOptions()
> 		retryOpts.Closer = m.clientStopper.ShouldQuiesce()
> ```

remove?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-:-KM_VEdt3KbPKfx6IzTu:452976905)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: deflake TestStoreMetrics

This was a tough one. Several problems were addressed, all variations on the
same theme:

- DistSenders in multiTestContext use a shared global stopper, but they
  may be called on goroutines which belong to a Store-level task. If that
  Store wants to quiesce and the DistSender can't finish its task because
  that same Store is already in quiescing mode, deadlocks occurred.
  The unfortunate solution is plugging in a channel which draws from two
  Stoppers, one of which may be quiesced and replaced multiple times.
- Additional deadlocks were caused due to multiTestContext's transport,
  which acquired a read lock that was formerly held in write mode throughout
  mtc.stopStore() (circumvented by dropping the lock there while quiescing).
- verifyStats was stopping individual Stores to perform computations without
  moving parts. Stopping individual Stores is tough when their tasks may be
  stuck on other Stores but can't complete while their own Store is already
  quiescing. Instead, verifyStats stops *all stores* simultaneously, regardless
  of which Store is actively being investigated.

Prior to these changes, failed in a few hundred to a few thousand iters
(depending on how many of the above were partially addressed):

```
$ make stressrace PKG=./storage TESTS=TestStoreMetrics TESTTIMEOUT=10s STRESSFLAGS='-maxfails 1 -stderr -p 128 -timeout 15m'
15784 runs so far, 0 failures, over 8m0s
```

Fixes #7678.

tschottdorf commented:
Review status: 2 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/client_metrics_test.go, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KM_UWcq0yBM0rsQmUDI:-KMa4GEE4X_rh9Q1e5n1:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_metrics_test.go#L82)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> danger zone
</details>
?

---

*[storage/client_metrics_test.go, line 102 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KM_Uob8idq49SNKutxO:-KMa58MM0J0CFHL1htGu:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_metrics_test.go#L102)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> since you're here, uncapitalize these
</details>
Done.

---

*[storage/client_metrics_test.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KM_CuY6NanDiHmTk7L_:-KMa5FKP9dHzk8x7aaou:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_metrics_test.go#L132)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this comment is truly worthless
</details>
Done.

---

*[storage/client_test.go, line 544 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KM_V4BXBwxUYPjgqM5-:-KMa5R-1eSdicVSBsN1g:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L544)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> remove?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 5 files at r1, 2 of 3 files at r2.
Review status: 4 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[storage/client_metrics_test.go, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KM_UWcq0yBM0rsQmUDI:-KMaCbqBxAQYJAmHVW1a:207806651) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_metrics_test.go#L82)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> ?
</details>
Nothing, this just looks like a race condition to me, but I guess that's what we're dealing with here.

---

*[storage/client_test.go, line 547 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMaIVCGE5eJxQSou7tU:-KMaIVCGE5eJxQSou7tV:-864553119) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L547)):*
> ```Go
> 		retryOpts.Closer = func() chan struct{} {
> 			ch := make(chan struct{})
> 			// Feed the channel periodically as long as our "real" stopper
> ```

this comment doesn't help me understand why it's ok to "feed" the closer channel - this is not a semantic we've ever used before, and it's not clear what's being achieved by doing this (there's no way to control who's on the other side of this "feeding").

---

*[storage/client_test.go, line 563 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMaHvm5usiIsn-s0XzQ:-KMaHvm5usiIsn-s0XzR:-662310894) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L563)):*
> ```Go
> 						return true
> 					}
> 					return false
> ```

nit: more readable in the `ch <- struct{}{}` case

---

*[storage/client_test.go, line 571 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMaID1kOrVdauczyN1P:-KMaID1kOrVdauczyN1Q:-481688185) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L571)):*
> ```Go
> 					for grabbedStopper != nil {
> 						m.mu.RLock()
> 						if len(m.stoppers) <= idx {
> ```

how come this check is needed?

---

*[storage/client_test.go, line 683 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMaGOXPm_Njr8l8iVD9:-KMaGOXPm_Njr8l8iVDA:-1865903358) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L683)):*
> ```Go
> func (m *multiTestContext) stopStore(i int) {
> 	m.mu.Lock()
> 	// Stopping when multiple stoppers (which are not aware of each other) is
> ```

grammar

---

*[storage/client_test.go, line 688 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMaG_J5StJ0dZ39Eu0L:-KMaG_J7miBSUXDB2eRM:-1150006953) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L688)):*
> ```Go
> 	// it's already in a task, so if we simply grabbed a write lock here while
> 	// stopping we could deadlock (see #7678).
> 	// So we initiate stopping under a write lock, and then release the lock
> ```

So we initiate quiescing under a write lock, and then release the lock during stopping.

---

*[storage/client_test.go, line 692 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMaGVu8Imv9u5Li6f_y:-KMaGVu8Imv9u5Li6f_z:-1922484539) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L692)):*
> ```Go
> 	stopper := m.stoppers[i]
> 	m.stoppers[i] = nil
> 	go func() {
> ```

`go stopper.Quiesce()`?

---

*[storage/client_test.go, line 700 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMaH3UraeR0PIyRjZZU:-KMaH3UraeR0PIyRjZZV:-1017772855) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L700)):*
> ```Go
> 
> 	m.mu.Lock()
> 	defer m.mu.Unlock()
> ```

nit: this defer looks awkward now, i'd just move it down

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 2 of 5 files at r1, 1 of 3 files at r2.
Review status: 4 of 5 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[storage/client_test.go, line 544 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMbiEqj0gqtyURE3R0B:-KMbiEqj0gqtyURE3R0C:564209831) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f318393127a947ee816780cef63f8d8ca70e180/storage/client_test.go#L544)):*
> ```Go
> 	if len(m.dbs) <= idx {
> 		retryOpts := base.DefaultRetryOptions()
> 		retryOpts.Closer = func() chan struct{} {
> ```

clientStopper is kind of an anachronism - prior to #4926 there was only one distSender so it got its own stopper. Now that each store gets its own distSender, those distSenders should probably use the store's stopper as well. I'm not sure what pulling on that thread would do, but it might be a way around this extra channel and goroutine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 4 of 5 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[storage/client_test.go, line 547 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMaIVCGE5eJxQSou7tU:-KMcGWVnxC30Mc7rTQWW:891630020) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L547)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this comment doesn't help me understand why it's ok to "feed" the closer channel - this is not a semantic we've ever used before, and it's not clear what's being achieved by doing this (there's no way to control who's on the other side of this "feeding").
</details>
This simulates a channel which is closed when either of two parent channels is closed (or nil). Horrible stuff and I hope that with Ben's comment above addressed, everything else remains stable and we can forget about this part.

---

*[storage/client_test.go, line 688 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMaG_J5StJ0dZ39Eu0L:-KMcGxdSJ1oCpkxHqZz9:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L688)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> So we initiate quiescing under a write lock, and then release the lock during stopping.
</details>
Done.

---

*[storage/client_test.go, line 692 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMaGVu8Imv9u5Li6f_y:-KMcH-AqLDLIZm8rjgxR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L692)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `go stopper.Quiesce()`?
</details>
Done.

---

*[storage/client_test.go, line 700 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMaH3UraeR0PIyRjZZU:-KMcH1grvCIWWVWf74Vc:-413379104) ([raw file](https://github.com/cockroachdb/cockroach/blob/46af5e32671d6e8177d510a294e264807b850414/storage/client_test.go#L700)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: this defer looks awkward now, i'd just move it down
</details>
`RemoveStore` could panic (or at least I don't want to assume it can't).

---

*[storage/client_test.go, line 544 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMbiEqj0gqtyURE3R0B:-KMc6hYf-XpyjwUabkgd:-276090865) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f318393127a947ee816780cef63f8d8ca70e180/storage/client_test.go#L544)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> clientStopper is kind of an anachronism - prior to #4926 there was only one distSender so it got its own stopper. Now that each store gets its own distSender, those distSenders should probably use the store's stopper as well. I'm not sure what pulling on that thread would do, but it might be a way around this extra channel and goroutine.
</details>
Good call, PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: remove clientStopper from multiTestContext

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/client_test.go, line 486 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMcjCpZxFWIPfPvCfmy:-KMcjCp_df8OYBYNvIHn:-2143196897) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d5c500462c46304e289f5e2e18d12223a640778/storage/client_test.go#L486)):*
> ```Go
> 
> func (m *multiTestContext) populateDB(idx int, stopper *stop.Stopper) {
> 	if len(m.dbs) <= idx {
> ```

this pattern seems valid only when `len(m.dbs) == idx`, and yet it's peppered throughout this file =/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: refactor index usage in multiTestContext

tschottdorf commented:
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/client_test.go, line 486 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMcjCpZxFWIPfPvCfmy:-KMco7g3v5PKH9Qhv0TB:676965300) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d5c500462c46304e289f5e2e18d12223a640778/storage/client_test.go#L486)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this pattern seems valid only when `len(m.dbs) == idx`, and yet it's peppered throughout this file =/
</details>
PTAL

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-:-KMcq45l5DUYGBqa-jJR:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/client_test.go, line 486 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMcjCpZxFWIPfPvCfmy:-KMcryrq-6XuQ0DpLCWB:-1889738827) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d5c500462c46304e289f5e2e18d12223a640778/storage/client_test.go#L486)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> PTAL
</details>
Have to look at this again actually. All of this opportunistic appending was done for a reason as some tests inject some of the items prior to the test starting (store context, clocks, ...).
I'm going to put in some assertions that make sure we only override what can be overridden.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: guard against invalid init in multiTestContext

tschottdorf commented:
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/client_test.go, line 486 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809#-KMcjCpZxFWIPfPvCfmy:-KMcsrV_iLSCoE2mC8On:1845135228) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d5c500462c46304e289f5e2e18d12223a640778/storage/client_test.go#L486)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Have to look at this again actually. All of this opportunistic appending was done for a reason as some tests inject some of the items prior to the test starting (store context, clocks, ...).
> I'm going to put in some assertions that make sure we only override what can be overridden.
</details>
Ok, good to go now with last commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7809)*
<!-- Sent from Reviewable.io -->
